File: Languages/cpp/algorithms-dsa/patterns/heaps.cpp
/**
 * Pattern: Heaps / Priority Queues
 *
 * Key techniques covered:
 * 1. Top-K Elements (max-heap / min-heap of size K)
 * 2. Merge K Sorted Lists (min-heap)
 * 3. Kth Largest Element (min-heap of size K)
 * 4. Running Median (two heaps: max-heap for lower half, min-heap for upper half)
 *
 * Core insight: priority_queue gives O(log n) insert/extract.
 * - Max-heap (default): priority_queue<int>
 * - Min-heap: priority_queue<int, vector<int>, greater<int>>
 *
 * Compile: g++ -std=c++17 -o test heaps.cpp && ./test
 */
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <cassert>
using namespace std;

// ─────────────────────────────────────────────────────────────────────────────
// 1. Top-K Frequent Elements
//    Strategy: Use a min-heap of size K to keep the K largest frequencies.
//    Time: O(n log k), Space: O(n)
// ─────────────────────────────────────────────────────────────────────────────
#include <unordered_map>

vector<int> topKFrequent(vector<int>& nums, int k) {
    // Count frequencies
    unordered_map<int, int> freq;
    for (int n : nums) freq[n]++;

    // Min-heap of (frequency, element) keeps K most frequent
    using P = pair<int, int>;
    priority_queue<P, vector<P>, greater<P>> minHeap;

    for (auto& [num, count] : freq) {
        minHeap.push({count, num});
        if ((int)minHeap.size() > k) {
            minHeap.pop(); // Remove least frequent
        }
    }

    vector<int> result;
    while (!minHeap.empty()) {
        result.push_back(minHeap.top().second);
        minHeap.pop();
    }
    return result;
}

// ─────────────────────────────────────────────────────────────────────────────
// 2. Merge K Sorted Lists (simulated with vectors)
//    Strategy: Push the first element of each list into a min-heap.
//    Pop smallest, push next element from that list.
//    Time: O(N log K) where N = total elements, K = number of lists
// ─────────────────────────────────────────────────────────────────────────────
vector<int> mergeKSorted(vector<vector<int>>& lists) {
    // (value, list_index, element_index)
    using T = tuple<int, int, int>;
    priority_queue<T, vector<T>, greater<T>> minHeap;

    // Initialize heap with first element of each list
    for (int i = 0; i < (int)lists.size(); i++) {
        if (!lists[i].empty()) {
            minHeap.push({lists[i][0], i, 0});
        }
    }

    vector<int> result;
    while (!minHeap.empty()) {
        auto [val, li, ei] = minHeap.top();
        minHeap.pop();
        result.push_back(val);

        // If there's a next element in the same list, push it
        if (ei + 1 < (int)lists[li].size()) {
            minHeap.push({lists[li][ei + 1], li, ei + 1});
        }
    }
    return result;
}

// ─────────────────────────────────────────────────────────────────────────────
// 3. Kth Largest Element
//    Strategy: Maintain a min-heap of size K.
//    After processing all elements, the top is the Kth largest.
//    Time: O(n log k), Space: O(k)
// ─────────────────────────────────────────────────────────────────────────────
int kthLargest(vector<int>& nums, int k) {
    priority_queue<int, vector<int>, greater<int>> minHeap;

    for (int n : nums) {
        minHeap.push(n);
        if ((int)minHeap.size() > k) {
            minHeap.pop(); // Remove smallest, keeping K largest
        }
    }
    return minHeap.top();
}

// ─────────────────────────────────────────────────────────────────────────────
// 4. Running Median (Two Heaps)
//    Strategy:
//    - maxHeap: stores the smaller half (top = max of lower half)
//    - minHeap: stores the larger half (top = min of upper half)
//    - Balance: maxHeap.size() >= minHeap.size() (differ by at most 1)
//    Median = maxHeap.top() if odd count, else average of both tops.
//    Time per insert: O(log n), Space: O(n)
// ─────────────────────────────────────────────────────────────────────────────
class MedianFinder {
    priority_queue<int> maxHeap;                            // Lower half
    priority_queue<int, vector<int>, greater<int>> minHeap; // Upper half

public:
    void addNum(int num) {
        // Always add to maxHeap first
        maxHeap.push(num);

        // Ensure maxHeap's top <= minHeap's top
        if (!minHeap.empty() && maxHeap.top() > minHeap.top()) {
            minHeap.push(maxHeap.top());
            maxHeap.pop();
        }

        // Balance sizes: maxHeap can have at most 1 more than minHeap
        if ((int)maxHeap.size() > (int)minHeap.size() + 1) {
            minHeap.push(maxHeap.top());
            maxHeap.pop();
        } else if ((int)minHeap.size() > (int)maxHeap.size()) {
            maxHeap.push(minHeap.top());
            minHeap.pop();
        }
    }

    double findMedian() {
        if (maxHeap.size() > minHeap.size()) {
            return maxHeap.top();
        }
        return (maxHeap.top() + minHeap.top()) / 2.0;
    }
};

// ─────────────────────────────────────────────────────────────────────────────
int main() {
    // Test 1: Top-K Frequent
    {
        vector<int> nums = {1, 1, 1, 2, 2, 3};
        auto result = topKFrequent(nums, 2);
        sort(result.begin(), result.end());
        assert(result == vector<int>({1, 2}));
    }

    // Test 2: Merge K Sorted Lists
    {
        vector<vector<int>> lists = {{1, 4, 5}, {1, 3, 4}, {2, 6}};
        auto merged = mergeKSorted(lists);
        assert(merged == vector<int>({1, 1, 2, 3, 4, 4, 5, 6}));
    }

    // Test 3: Kth Largest
    {
        vector<int> nums = {3, 2, 1, 5, 6, 4};
        assert(kthLargest(nums, 2) == 5);

        vector<int> nums2 = {3, 2, 3, 1, 2, 4, 5, 5, 6};
        assert(kthLargest(nums2, 4) == 4);
    }

    // Test 4: Running Median
    {
        MedianFinder mf;
        mf.addNum(1);
        assert(mf.findMedian() == 1.0);
        mf.addNum(2);
        assert(mf.findMedian() == 1.5);
        mf.addNum(3);
        assert(mf.findMedian() == 2.0);
        mf.addNum(4);
        assert(mf.findMedian() == 2.5);
        mf.addNum(5);
        assert(mf.findMedian() == 3.0);
    }

    cout << "All heap pattern tests passed!" << endl;
    return 0;
}

File: Languages/java/algorithms-dsa/patterns/Heaps.java
import java.util.*;

/**
 * Heap / Priority Queue Patterns
 * 
 * Key insight: Heaps give O(log n) insert and O(1) access to min/max element.
 * Use a min-heap when you need the smallest, max-heap when you need the largest.
 * 
 * Common patterns:
 * 1. Top-K elements: Use a min-heap of size K (keeps K largest)
 * 2. Merge K sorted: Use a min-heap to always pick the smallest next element
 * 3. Kth largest: Maintain a min-heap of size K; root is the answer
 * 4. Running median: Use two heaps (max-heap for lower half, min-heap for upper half)
 * 
 * Time complexities:
 * - Top-K: O(n log k)
 * - Merge K sorted: O(n log k) where n = total elements, k = number of lists
 * - Running median: O(log n) per insert, O(1) per findMedian
 */
public class Heaps {

    // ==================== Top-K Frequent Elements ====================
    /**
     * Find the k most frequent elements in an array.
     * Strategy: Count frequencies, then use a min-heap of size k.
     * The heap evicts the least frequent, leaving the k most frequent.
     * 
     * Time: O(n log k), Space: O(n)
     */
    public static int[] topKFrequent(int[] nums, int k) {
        // Step 1: Count frequencies
        Map<Integer, Integer> freq = new HashMap<>();
        for (int num : nums) {
            freq.put(num, freq.getOrDefault(num, 0) + 1);
        }

        // Step 2: Min-heap by frequency, keep size k
        PriorityQueue<Integer> minHeap = new PriorityQueue<>(
            (a, b) -> freq.get(a) - freq.get(b)
        );

        for (int num : freq.keySet()) {
            minHeap.offer(num);
            if (minHeap.size() > k) {
                minHeap.poll(); // Remove least frequent
            }
        }

        // Step 3: Extract results
        int[] result = new int[k];
        for (int i = 0; i < k; i++) {
            result[i] = minHeap.poll();
        }
        return result;
    }

    // ==================== Merge K Sorted Lists ====================
    /**
     * Merge k sorted arrays into one sorted array.
     * Strategy: Use a min-heap holding one element from each list.
     * Always extract the minimum and advance that list's pointer.
     * 
     * Time: O(n log k), Space: O(k) for the heap
     */
    public static int[] mergeKSorted(int[][] lists) {
        // Min-heap stores: [value, listIndex, elementIndex]
        PriorityQueue<int[]> minHeap = new PriorityQueue<>(
            (a, b) -> a[0] - b[0]
        );

        // Initialize: add first element of each list
        int totalSize = 0;
        for (int i = 0; i < lists.length; i++) {
            if (lists[i].length > 0) {
                minHeap.offer(new int[]{lists[i][0], i, 0});
                totalSize += lists[i].length;
            }
        }

        int[] result = new int[totalSize];
        int idx = 0;

        while (!minHeap.isEmpty()) {
            int[] curr = minHeap.poll();
            result[idx++] = curr[0];

            int listIdx = curr[1];
            int elemIdx = curr[2] + 1;

            // If there's a next element in that list, add it
            if (elemIdx < lists[listIdx].length) {
                minHeap.offer(new int[]{lists[listIdx][elemIdx], listIdx, elemIdx});
            }
        }

        return result;
    }

    // ==================== Kth Largest Element ====================
    /**
     * Find the kth largest element in an unsorted array.
     * Strategy: Maintain a min-heap of size k.
     * After processing all elements, the root is the kth largest.
     * 
     * Time: O(n log k), Space: O(k)
     */
    public static int kthLargest(int[] nums, int k) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();

        for (int num : nums) {
            minHeap.offer(num);
            if (minHeap.size() > k) {
                minHeap.poll(); // Remove smallest, keep k largest
            }
        }

        return minHeap.peek(); // Root = kth largest
    }

    // ==================== Running Median (MedianFinder) ====================
    /**
     * MedianFinder maintains a running median using two heaps:
     * - maxHeap: stores the smaller half (gives quick access to max of lower half)
     * - minHeap: stores the larger half (gives quick access to min of upper half)
     * 
     * Invariant: maxHeap.size() == minHeap.size() or maxHeap.size() == minHeap.size() + 1
     * Median = maxHeap.peek() if odd count, else average of both peeks.
     */
    static class MedianFinder {
        private PriorityQueue<Integer> maxHeap; // Lower half
        private PriorityQueue<Integer> minHeap; // Upper half

        public MedianFinder() {
            maxHeap = new PriorityQueue<>(Collections.reverseOrder());
            minHeap = new PriorityQueue<>();
        }

        public void addNum(int num) {
            // Always add to maxHeap first
            maxHeap.offer(num);

            // Balance: ensure maxHeap's max <= minHeap's min
            if (!minHeap.isEmpty() && maxHeap.peek() > minHeap.peek()) {
                minHeap.offer(maxHeap.poll());
            }

            // Balance sizes: maxHeap can have at most 1 more than minHeap
            if (maxHeap.size() > minHeap.size() + 1) {
                minHeap.offer(maxHeap.poll());
            } else if (minHeap.size() > maxHeap.size()) {
                maxHeap.offer(minHeap.poll());
            }
        }

        public double findMedian() {
            if (maxHeap.size() > minHeap.size()) {
                return maxHeap.peek();
            }
            return (maxHeap.peek() + minHeap.peek()) / 2.0;
        }
    }

    // ==================== Tests ====================
    public static void main(String[] args) {
        // Test Top-K Frequent
        int[] topK = topKFrequent(new int[]{1, 1, 1, 2, 2, 3}, 2);
        Arrays.sort(topK);
        assert Arrays.equals(topK, new int[]{1, 2}) : "Top-K test 1 failed";

        int[] topK2 = topKFrequent(new int[]{1}, 1);
        assert topK2[0] == 1 : "Top-K test 2 failed";

        // Test Merge K Sorted
        int[][] lists = {{1, 4, 7}, {2, 5, 8}, {3, 6, 9}};
        int[] merged = mergeKSorted(lists);
        assert Arrays.equals(merged, new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9})
            : "Merge K sorted test failed";

        int[][] lists2 = {{1, 3, 5}, {2, 4, 6}, {0, 7}};
        int[] merged2 = mergeKSorted(lists2);
        assert Arrays.equals(merged2, new int[]{0, 1, 2, 3, 4, 5, 6, 7})
            : "Merge K sorted test 2 failed";

        // Test Kth Largest
        assert kthLargest(new int[]{3, 2, 1, 5, 6, 4}, 2) == 5
            : "Kth largest test 1 failed";
        assert kthLargest(new int[]{3, 2, 3, 1, 2, 4, 5, 5, 6}, 4) == 4
            : "Kth largest test 2 failed";

        // Test MedianFinder
        MedianFinder mf = new MedianFinder();
        mf.addNum(1);
        assert mf.findMedian() == 1.0 : "Median test 1 failed";
        mf.addNum(2);
        assert mf.findMedian() == 1.5 : "Median test 2 failed";
        mf.addNum(3);
        assert mf.findMedian() == 2.0 : "Median test 3 failed";
        mf.addNum(4);
        assert mf.findMedian() == 2.5 : "Median test 4 failed";
        mf.addNum(5);
        assert mf.findMedian() == 3.0 : "Median test 5 failed";

        System.out.println("All heap pattern tests passed!");
    }
}
